### PR TITLE
Docs for citation functionality

### PR DESCRIPTION
**Pull request by willGraham01**
Provides documentation for the `cite-brainglobe` tool in https://github.com/brainglobe/brainglobe-utils/pull/28.

Also related to https://github.com/brainglobe/BrainGlobe/issues/61.

### CODE REVIEW

File: docs/source/documentation/brainglobe-utils/citation-module.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+```
+
+The `-l` (`--list`) option will provide you with a list of citation formats that the tool supports, and the list of tools that the program is aware of.
+Currently supported citation formats are:
+
+- Bibtex (`*.tex`), use `--format bibtex` to request this citation type.
+- Text (`*.txt`), use `--format text` to request this citation type. This option is mainly for when you want to generate a citation you can copy/paste into a bibliography, or an acknowledgements section.
+
+The `-s` (`--software-citations`) option will prioritise citing BrainGlobe tool _software_ - that is, the source code or program - rather than the article or journal entry that provides the theoretical basis for the tool or algorithm.
+By default we expect users to prefer citing the article, however if you specifically want to credit the software or tool implementation - in cases where you have made a contribution to the source code for example - you can use this option.
```
Comment (alessandrofelder) [inline docs/source/documentation/brainglobe-utils/citation-module.md:101]: I wonder whether we should encourage citation of both :thinking: maybe one for the dev meeting to discuss?

**Review by alessandrofelder — COMMENTED** [review]
Happy with this in principle, but have some minor discussion points.
Probably worth @adamltyson having a look at this too!

File: docs/source/documentation/brainglobe-utils/citation-module.md (hunk 1 above)
Comment (willGraham01) [reply docs/source/documentation/brainglobe-utils/citation-module.md:101]: Could just change the default behaviour to "all", which reads any and all references provided.

Then add a command-line switch for "only articles", "only software", etc?

File: docs/source/documentation/brainglobe-utils/citation-module.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    title = "BrainGlobe",
+    url = "https://brainglobe.info/",
+    year = "2024",
+    abstract = "The BrainGlobe Initiative exists to facilitate the development of interoperable Python-based tools for computational neuroanatomy.",
+    license = "BSD-3-Clause",
+}
+```
+
+By default, citations are printed to the terminal screen (`stdout`), which you can then copy where you need.
+The default output format for citations is `bibtex`.
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply docs/source/documentation/brainglobe-utils/citation-module.md:63]: No worries - this can be easily switched to text default (just one parameter needs changing)

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    "bg-atlasapi",
+    [
+        "bg_atlasapi",
+        "BrainGlobe AtlasAPI",
+        "BrainGlobe AtlasAPI",
+        "AtlasAPI",
+        "Atlas API",
+    ],
+)
+```
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply docs/source/community/developers/repositories/brainglobe-utils/index.md:71]: Yeah we should and can do the same example with brainglobe-meta/utils now - bg-atlasapi was the one I was testing against at the time because it's CITATION.cff is now on main.

File: docs/source/documentation/brainglobe-utils/citation-module.md (hunk 1 above)
Comment (alessandrofelder) [reply docs/source/documentation/brainglobe-utils/citation-module.md:101]: I think the "all" default would be my preference, but I'll let @adamltyson chime in when he has a chance to look at this.

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
+# `brainglobe-utils`
```
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:1]: For consistency with the other toctree headings
```suggestion
# brainglobe-utils
```

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+`brainglobe-utils` is a collection of functions that can be re-used across multiple repositories: for BrainGlobe tools, plugins, or analyses (in the case of `brainglobe-workflows`).
+It underpins a number of our user-facing tools like `brainreg` and `cellfinder`, as well as our common interfaces to napari in `brainglobe-napari-io`, for example.
+The main feature it provides to users is the [`cite-brainglobe`](#cite-brainglobe-and-the-citation-submodule) command-line program.
+
+## `cite-brainglobe` and the `citation` submodule
+
+The `cite-brainglobe` command-line program is provided by the `citation` submodule, with the command-line wrapper and main function themselves being available in `citation.cite:cli` and `citation.cite:cite` respectively.
+
+At the top level, the workflow for the citation tool is as follows:
```
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:11]:
```suggestion
At the top level, the workflow for the citation tool works as follows:
```

File: docs/source/community/developers/repositories/brainglobe-utils/index.md
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:1]: Is it worth having some links to the codebase here? I find it a bit hard to parse descriptions of code without seeing the code itself.

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+Finally, in your class you'll need to overwrite the `generate_ref_string` method to produce the citation string from the metadata information you specified.
+
+Then, head to the `citation.cite` module and add your new format as an option to the command-line interface (`citation.cite.cli`) and backend function (`citation.cite.cite`).
+
+### `citation.repositories`
+
+This module contains the `Repository` class, a function to find the repositories that are referred to in a list of tool names, and the static instances of all the BrainGlobe repositories that we provide a reference for via `cite-brainglobe`.
+
+The `Repository` class is just a convenient container for all the information pertaining to one particular BrainGlobe tool or repository.
+The class itself just holds any information we need and some useful actions on that information - like providing the URL for the GitHub repo, storing the alternative names for the tool that is held there, etc.
```
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:47]:
```suggestion
The class itself just holds any information we need and some useful actions on that information, such as providing the URL for the GitHub repo, storing the alternative names for the tool that is held there, etc.
```

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+The static instances of this class are instantiated in this module too - by convention, the variable name should match the repository's name on GitHub.
+Each instance is created using the syntax:
+
+```python
+bg_tool = Repository("bg-tool", ["list", "of", "alternative", "or", "informal", "names"])
```
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:52]: Just to follow the new naming convention. 
```suggestion
brainglobe_tool = Repository("brainglobe-tool", ["list", "of", "alternative", "or", "informal", "names"])
```

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+```
+
+This defines a repository that `cite-brainglobe` expects to be called `bg-tool`, under the `brainglobe` organisation on GitHub.
```
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:55]:
```suggestion
bg_tool = Repository("bg-tool", ["list", "of", "alternative", "or", "informal", "names"])
```

This defines a repository that `cite-brainglobe` expects to be called `brainglobe-tool`, under the `brainglobe` organisation on GitHub.
```

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    "brainglobe_utils",
+    [
+        "utilities",
+    ],
+)
+```
+
+creates a `Repository` object which, if the user asks `cite-brainglobe` to cite any one of `"brainglobe-utils"`, `"brainglobe_utils"`, `"brainglobe utils"`, `"utils"`, or `"utilities"`, the program will recognise as the `brainglobe-utils` package.
+
+The `all_citable_repositories` function is intended for imports in other areas of th `citation` submodule - it automatically detects the static `Repository` instances that we define in the `repositories` and returns them as a set.
```
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:75]:
```suggestion
The `all_citable_repositories` function is intended for imports in other areas of the `citation` submodule — it automatically detects the static `Repository` instances that we define in the `repositories` and returns them as a set.
```

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+The `unique_repositories_from_tools` function takes in a list of tool names or aliases (in particular, the list provided by the user on the command-line) and returns the unique repositories that need to be cited given this list.
+
+### `citation.fetch`
+
+This submodule contains two helper functions.
+The simplest is the `yaml_str_to_dict` function which takes a string containing the yaml-formatted content of a file and parses it into a Python `dict` - this is for use when retrieving files from the internet as opposed to loading them from disk.
```
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:82]:
```suggestion
The simplest is the `yaml_str_to_dict` function which takes a string containing the yaml-formatted content of a file and parses it into a Python `dict` — this is for use when retrieving files from the internet as opposed to loading them from disk.
```

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+The key properties of the `Format` class are the `required` and `optional` class variables - these are not set in the base class and are intended to be overwritten by classes that inherit from `Format`.
+Both of these class variables should be lists of strings, corresponding to the keys in a CITATION.cff format that the citation (described by the inheriting class) needs (respectively can make use of) when being produced.
+Existence of these keys in the CITATION.cff information is checked for in the `Format.__init__` class constructor to avoid repetition across submodules and catch errors.
+Optional keys are allowed to be missing - these can be set to be reported if so.
+The `Format` class also contains a class-wide function for parsing the `authors` key information into a string - `Format._prepare_authors_field` - which is again invoked on instantiation.
+Finally, `Format` provides a placeholder `generate_ref_string` function that should be overwritten by any inheriting classes - this is the method that will parse the data read in and produce the citation string.
```
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:97]:
```suggestion
Finally, `Format` provides a placeholder `generate_ref_string` function that should be overwritten by any inheriting classes — this is the method that will parse the data read in and produce the citation string.
```

File: docs/source/community/developers/repositories/brainglobe-utils/index.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+For example, the `TextCitation` format inherits from `Format` and requires the keys "author", "title" and "year" to produce a reference.
+It also provides a list of optional keys that the format can still make use of when producing the citation, but it does not require them.
+`TextCitation` itself does not need to define a constructor function since the inherited constructor from `Format` now suffices with the overwritten values for `TextCitation.required` and `TextCitation.optional`.
+`TextCitation` does implement `generate_ref_string` in order to overwrite the placeholder function defined by `Format` - this will dictate how the citation string is formatted and assembled.
```
Comment (adamltyson) [inline docs/source/community/developers/repositories/brainglobe-utils/index.md:102]:
```suggestion
`TextCitation` does implement `generate_ref_string` in order to overwrite the placeholder function defined by `Format` — this will dictate how the citation string is formatted and assembled.
```

File: docs/source/documentation/brainglobe-utils/citation-module.md (hunk 1 above)
Comment (adamltyson) [reply docs/source/documentation/brainglobe-utils/citation-module.md:101]: I would like all, but I think paper first is a more practical solution. Many journals have citation limits (:shrug:) and BG tools will be one of 100s of citations. We want to make it as simple as possible to cite us.

File: docs/source/documentation/brainglobe-utils/citation-module.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamltyson) [reply docs/source/documentation/brainglobe-utils/citation-module.md:63]: Yeah I think text should be the default. Even though some experimentalists know latex, very few journals accept it.

**Review by adamltyson — APPROVED** [review]
Looks very nice @willGraham01. I didn't go into too much detail because I'm not familiar with the citation PR myself. I've also very pedantically removed lots of rogue hyphens.

**Comment by alessandrofelder** [comment]
> Can this be merged now?

I guess so
